### ISSUE TITLE
C407 increases laziness which should be noted since it can change behaviour

### ISSUE THREAD

**Issue by merc1031** [adamchainz/flake8-comprehensions#247]
The following issues the C407 warning (as desired)
```
results = all([m.delete() for m in q.where_in(self._other_key, ids).get()])
```
however if you appease it and switch the code to 
```
results = all((m.delete() for m in q.where_in(self._other_key, ids).get()))
```
or
```
results = all(m.delete() for m in q.where_in(self._other_key, ids).get())
```
It causes a subtle bug due to the new laziness gained by the generator expression

This can be evidenced here
```
>>> all((print(i) for i in range(1, 10)))
1
False
>>> all((print(i) or True for i in range(1, 10)))
1
2
3
4
5
6
7
8
9
True
```

This warning should probably remind/warn (haha) the user that there is a risk in changing it to follow the warning if the generate has a side effecting function in it

**Comment by adamchainz** [comment]
Rule C407 has been a constant source of problems - 5 or 6 bugs in the changelog: https://github.com/adamchainz/flake8-comprehensions/blob/master/HISTORY.rst . I'm tempted to just remove it. I'd prefer this library gave 100% actionable advice rather than "sometimes this is a good idea" tweaks. What do you think?

**Comment by merc1031** [comment]
Yah, its tought.
It seems like a decent rule, but it can bite you pretty hard.
Maybe if it defaults to being ignored, and has a more strongly worder warning about its implications? That could be a good compromise.

in the haskell linter, hlint, there are a few lints that warn you that they either increase or decrease laziness.

**Comment by DylanYoung** [comment]
This subtlety also affects C412, eh?   At least a number of builtins aren't affected (sorted, sum,...etc.).

Maybe it should be split into two error codes? One safe one that's enabled by default for builtins that consume the generator and one unsafe one that isn't?

**Comment by smithdc1** [comment]
Happy new year @adamchainz 

I ran your tool against Django and it highlighted a few items. I tried to evidence that a generator was more performant, but surprisingly it is not, which I think you already knew based on comments you made against a Django PR. Anyway, I've put together a benchmark that provides an example of this [here](https://gist.github.com/smithdc1/7825842972afe05ad2f457e8b9fbc890) where a list comprehension is begin passed to `tuple()`.

Using my benchmark script, the generator version is ~50% (!) slower, and this scales to reasonably large values. I'm not sure the large data set times are particularly accurate as they are fairly large, but the general indication is ok. I'm using Python 3.9 to run these benchmarks. 

```
> python bench.py
.....................
list_comprehension small: Mean +- std dev: 16.8 us +- 0.3 us
.....................
generator small: Mean +- std dev: 25.8 us +- 0.6 us
.....................
list_comprehension large: Mean +- std dev: 580 ms +- 56 ms
.....................
generator large: Mean +- std dev: 921 ms +- 26 ms
```

**Comment by adamchainz** [comment]
Given the negative impacts on laziness and performance I dropped both rules in #312. Thanks for the input and benchmarks all.
